src/lightning/fabric/plugins/precision/fp8_transformer_engine.py

```diff
@@ -0,0 +1,181 @@
+# Copyright The Lightning AI team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from contextlib import contextmanager
+from typing import Any, Generator, Literal, Mapping, Optional, Tuple, Type, TYPE_CHECKING, Union
+
+import torch
+from lightning_utilities import apply_to_collection
+from lightning_utilities.core.imports import RequirementCache
+from torch import Tensor
+
+from lightning.fabric.plugins.precision.precision import Precision
+from lightning.fabric.plugins.precision.utils import _convert_fp_tensor
+from lightning.fabric.utilities.rank_zero import rank_zero_warn
+
+_TRANSFORMER_ENGINE_AVAILABLE = RequirementCache("transformer_engine")
+
+if TYPE_CHECKING and _TRANSFORMER_ENGINE_AVAILABLE:
+    from transformer_engine.common.recipe import DelayedScaling
+else:
+    DelayedScaling = None
+
+
+log = logging.getLogger(__name__)
+
+
+class Fp8TransformerEnginePrecision(Precision):
+    """Plugin for training with fp8 precision via nvidia's `Transformer Engine
+    <https://docs.nvidia.com/deeplearning/transformer-engine`__.
+
+    .. warning::  This is an :ref:`experimental <versioning:Experimental API>` feature.
+
+    Args:
+        dtype: The base dtype to use.
+        recipe: Recipe for the DelayedScaling
+            `configuration <https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/common.html#transform
+            er_engine.common.recipe.DelayedScaling`__. In dict format or the dataclass format.
+        replace_layers: Whether to replace ``Linear`` and ``LayerNorm`` layers automatically with their Transformer
+            Engine alternatives. Note that they don't subclass the torch equivalents so checks like
+            ``isinstance(l, torch.nn.Linear)`` will not pass.
+
+    .. note::
+
+        Support for FP8 in the linear layers with `precision='8-mixed'` is currently limited to tensors with shapes
+        where the dimensions are divisible by 8 and 16 respectively. You might want to add padding to your inputs to
+        conform to this restriction.
+    """
+
+    precision: Literal["8-mixed-transformer-engine"] = "8-mixed-transformer-engine"
+
+    def __init__(
+        self,
+        dtype: Optional[torch.dtype] = None,
+        recipe: Optional[Union[Mapping[str, Any], "DelayedScaling"]] = None,
+        replace_layers: bool = True,
```

Comment:
Shouldn't this be None?
Shouldn't explicitly setting it to True and already having manually replaced layers raise an error, as it would be a conflicting request.

Reply:
This would mean that when `init_module()` is used, `replace_layers` needs to be set to False if it was True previously. Additionally, we don't check whether layers have been replaced already in the `init_module` case because the model doesn't exist. I don't think this check is worth it.